### PR TITLE
Improve error handling for in-evaluable refs for discriminator application

### PR DESCRIPTION
**Pull request by sydney-runkle**
This was supposed to close https://github.com/pydantic/pydantic/issues/8271, but right now it just improves the associated error.

My current hunch is that we need to do an apply discriminators step at the end of json schema building, like we do for core schemas, if we want this to work 😭. Not sure how worth the investment that is, at the moment.

I'd also be open to more creative solutions for evaluating refs for which we have not yet generated the schemas. Probably a good conversation to have with @Viicos and @adriangb.

MRE, at the moment:

```py
from enum import Enum
from typing import Annotated, Union, Literal, List

from pydantic import BaseModel as BaseModel, Field, TypeAdapter
from pydantic.fields import FieldInfo
import json
from pydantic._internal._core_utils import pretty_print_core_schema

class ItemType(str, Enum):
    ITEM1 = 'item1'
    ITEM2 = 'item2'

class CreateItem1(BaseModel):
    item_type: Annotated[Literal[ItemType.ITEM1], Field(alias='type')]
    id: int

class CreateItem2(BaseModel):
    item_type: Annotated[Literal[ItemType.ITEM2], Field(alias='type')]
    id: int

class CreateObjectDto(BaseModel):
    id: int
    items: List[
        Annotated[
            Union[
                CreateItem1,
                CreateItem2,
            ],
            Field(discriminator='item_type'),
        ]
    ]

ta = TypeAdapter(
    Annotated[CreateObjectDto, FieldInfo(examples=[{'id': 1, 'items': [{'id': 3, 'type': 'ITEM1'}]}])]
)

print(json.dumps(ta.json_schema(), indent=2))
print(pretty_print_core_schema(ta.core_schema))
```

### CODE REVIEW

File: pydantic/json_schema.py [hunk 1]
```diff
@@ -466,7 +466,7 @@ def populate_defs(core_schema: CoreSchema, json_schema: JsonSchemaValue) -> Json
                 core_ref = CoreRef(core_schema['ref'])  # type: ignore[typeddict-item]
                 defs_ref, ref_json_schema = self.get_cache_defs_ref_schema(core_ref)
                 json_ref = JsonRef(ref_json_schema['$ref'])
-                self.json_to_defs_refs[json_ref] = defs_ref
+                # self.json_to_defs_refs[json_ref] = defs_ref
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:469]:
```suggestion
```

File: pydantic/json_schema.py [hunk 2]
```diff
@@ -466,7 +466,6 @@ def populate_defs(core_schema: CoreSchema, json_schema: JsonSchemaValue) -> Json
                 core_ref = CoreRef(core_schema['ref'])  # type: ignore[typeddict-item]
                 defs_ref, ref_json_schema = self.get_cache_defs_ref_schema(core_ref)
                 json_ref = JsonRef(ref_json_schema['$ref'])
-                self.json_to_defs_refs[json_ref] = defs_ref
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:469]: This is done in the `self.get_cache_defs_ref_schema(core_ref)` call, so no reason to do that here.

File: pydantic/json_schema.py [hunk 3]
```diff
@@ -1201,9 +1200,7 @@ def _extract_discriminator(
                     continue  # this means that the "alias" does not represent a field
                 alias_is_present_on_all_choices = True
                 for choice in one_of_choices:
-                    while '$ref' in choice:
-                        assert isinstance(choice['$ref'], str)
-                        choice = self.get_schema_from_definitions(JsonRef(choice['$ref'])) or {}
+                    choice = self.resolve_schema_to_update(choice)
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:1203]: This function handles the above removed logic more gracefully, with an intuitive error.

**Review by Viicos — APPROVED** [review]
I'll try to do a deep dive in the JSON Schema generation process to see what this is about, I'm guessing it is indeed pretty similar to the core schema gen process

**Comment by sydney-runkle** [comment]
Note from my conversation with @adriangb - we could alternatively do a first pass to collect all defs

**Pull request merged**